frontend/components/FieldColor/FieldColorEdit.jsx

```diff
@@ -0,0 +1,262 @@
+     * new value of the field.
+     */
+    onError: proptypes.string,
+
+    /**
+     * Whether the field is required.
+     */
+    required: proptypes.bool,
+
+    /**
+     * The field schema.
+     */
+    schema: proptypes.object,
+
+    /**
+     * The field value.
+     */
+    value: proptypes.bool
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state.pickerVisible = false
+
+    this.picker = false
+    this.pickerEventHandler = this.handlePickerClick.bind(this, true)
+    this.pickerOutsideEventHandler = this.handlePickerClick.bind(this, false)
+    this.losingFocusTimeout = null
+    this.state.hasFocus = false
+  }
+
+  render() {
+    let {
+      comment,
+      config,
+      displayName,
+      error, 
+      name,
+      required,
+      schema, 
+      value
+    } = this.props
+    const {hasFocus} = this.state
+    const {pickerVisible} = this.state
+    const publishBlock = schema.publish || {}
+
+    return (
+      <Label
+        className={styles.label}
+        error={error}
+        errorMessage={typeof error === 'string' ? error : null}
+        hasFocus={hasFocus}
+        label={displayName}
+        comment={comment || (required && 'Required') || (publishBlock.readonly && 'Read only')}
+      > 
+        <TextInput
+          name={name}
+          onBlur={this.handleFocus.bind(this, false)}
+          onChange={this.handleChange.bind(this)}
+          onFocus={this.handleFocus.bind(this, true)}
+          readonly={publishBlock.readonly === true}
+          type="text"
+          value={value}
+        />
+
+        <div 
+          class={styles.swatch}
+          style={value ? `background-color:#${value}` : ''}
+        />
+
+        {pickerVisible &&
+          <div ref={this.handlePickerRef.bind(this)}>
+            <ColorPicker
+              className={styles.picker}
+              color={value}
+              onChange={this.handlePickerChange.bind(this)}
+              format={publishBlock.format || 'hex'}
+            />
+          </div>
+        }
+      </Label>
+    )
+  }
+
+  componentDidMount() {
+    window.addEventListener('click', this.pickerOutsideEventHandler)
+  }
+
+  componentWillUnmount() {
+    if (this.picker) {
+      this.picker.removeEventListener('click', this.pickerOutsideEventHandler)
+    }
+
+    clearTimeout(this.losingFocusTimeout)
+  }
+
+  handleChange(event) {
+    const {
+      config,
+      name,
+      onChange,
+      schema,
+      value
+    } = this.props
+
+    if (typeof onChange === 'function') {
+      onChange.call(this, name, event.target.value)
+    }
+  }
+
+  handleFocus(hasFocus) {
+    const {pickerVisible} = this.state
+    const {schema} = this.props
+
+    const publishBlock = schema.publish || {}
+
+    // Return from focus event so picker doesn't display
+    if (publishBlock.readonly === true) {
+      return
+    }
+
+    this.hasFocus = hasFocus
+
+    this.setState({
```

Comment:
You're saving `hasFocus` both on an instance variable (L208) and in the component state. Is there a reason why you need both?